tools/azure-pipelines/build_properties.sh

```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+#
+# Returns 0 if the change is a documentation-only pull request
+#
+function is_docs_only_pullrequest() {
+	# check if it is a pull request:
+	if [[ ! $BUILD_SOURCEBRANCHNAME == ci_* ]] ; then
+		return 1
+	fi
+	PR_ID=`echo "$BUILD_SOURCEBRANCHNAME" | cut -f2 -d_`
+	if ! [[ "$PR_ID" =~ ^[0-9]+$ ]] ; then
+		echo "ERROR: Extracted PR_ID is not a number, but this: '$PR_ID'"
+	 	return 1
+	fi
+	# check if it is docs only pull request:
+	CHANGES=`curl --silent "https://api.github.com/repos/apache/flink/pulls/$PR_ID/files" | jq -r ".[].filename"`
```

Comment:
This can fail in surprising ways if the GitHub view of the machines that this script and ci-bot run on differ.
E.g., PR contains commit only changing docs, another commit is added that changes something else -> ci-bot mirrors PR -> the CI machines sees the old state, before the new commit was added.
This doesn't even seem that unlikely given that the machines are located in different continents.

Reply:
I agree that there's a small risk for such situations, but I believe it is unlikely enough that we can risk it (unless you know a simple solution).
For the scenario you are describing, even if the timing of ci-bot is optimal, there's probably a delay of at least 2 minutes between "commit has been added" and "azure checks list of files". (Factors adding delay: ci-bot does not permanently check for new changes; pushing the branch takes some time; azure doesn't immediately kick off the build; the build doesn't immediately check the list of files).

A fairly common scenario might be that the e2e build gets queued, and the "get list of files" happens hours later. But the worst that can happen here is that we are wasting 4 hours of CI time for a docs only change (because files were added later).

I believe this change is a "best effort" that saves us some CI time in 99% of the cases where it makes sense.

Reply:
> But the worst that can happen here is that we are wasting 4 hours of CI time for a docs only change (because files were added later).

This is a case I also wouldn't care about; but it's not the worst case: a build that changes something outside docs/ does not get tested.

To solve this we just need to match the commit hashes.
https://api.github.com/repos/apache/flink/pulls/12708 contains the head commit of the PR, and number of commits.
Compare the hash against the head commit of the ci branch.
If they match, `git diff --name-only head..head~${num_commits}`, continue as usual.
If the hashes don't match, just run the tests.

Reply:
Thanks, I'll do that.